cirq/circuits/circuit_test.py

```diff
@@ -3572,3 +3572,91 @@ def test_transform_qubits():
     assert c.transform_qubits(lambda q: q).device is cg.Foxtail
     assert c.transform_qubits(lambda q: q, new_device=cg.Bristlecone
                              ).device is cg.Bristlecone
+
+
+def test_indexing_by_pair():
+    # 0: ───H───@───X───@───
+    #           │       │
+    # 1: ───────H───@───@───
+    #               │   │
+    # 2: ───────────H───X───
+    q = cirq.LineQubit.range(3)
+    c = cirq.Circuit([
+        cirq.H(q[0]),
+        cirq.H(q[1]).controlled_by(q[0]),
+        cirq.H(q[2]).controlled_by(q[1]),
+        cirq.X(q[0]),
+        cirq.CCNOT(*q),
+    ])
+
+    # Indexing by single moment and qubit.
+    assert c[0, q[0]] == c[0][q[0]] == cirq.H(q[0])
+    assert c[1, q[0]] == c[1, q[1]] == cirq.H(q[1]).controlled_by(q[0])
+    assert c[2, q[0]] == c[2][q[0]] == cirq.X(q[0])
+    assert c[2, q[1]] == c[2, q[2]] == cirq.H(q[2]).controlled_by(q[1])
+    assert c[3, q[0]] == c[3, q[1]] == c[3, q[2]] == cirq.CCNOT(*q)
+
+    # Indexing by moment and qubit - throws if there is no operation.
+    with pytest.raises(KeyError, match="Moment doesn't act on given qubit"):
+        _ = c[0, q[1]]
+
+    # Indexing by single moment and multiple qubits.
+    assert c[0, q] == c[0]
+    assert c[1, q] == c[1]
+    assert c[2, q] == c[2]
+    assert c[3, q] == c[3]
+    assert c[0, q[0:2]] == c[0]
+    assert c[0, q[1:3]] == cirq.Moment([])
+    assert c[1, q[1:2]] == c[1]
+    assert c[2, [q[0]]] == cirq.Moment([cirq.X(q[0])])
+    assert c[2, q[1:3]] == cirq.Moment([cirq.H(q[2]).controlled_by(q[1])])
+    assert c[2, q[0:2]] == c[2]
+
+    # Indexing by single qubit.
+    assert c[:, q[0]] == cirq.Circuit([
+        cirq.H(q[0]),
+        cirq.H(q[1]).controlled_by(q[0]),
+        cirq.X(q[0]),
+        cirq.CCNOT(*q),
+    ])
+    assert c[:, q[1]] == cirq.Circuit([
+        cirq.H(q[1]).controlled_by(q[0]),
+        cirq.H(q[2]).controlled_by(q[1]),
+        cirq.CCNOT(*q),
+    ])
+    assert c[:, q[2]] == cirq.Circuit([
+        cirq.H(q[2]).controlled_by(q[1]),
+        cirq.CCNOT(*q),
+    ])
```

Comment:
These should have empty moments so the kept operations are at the same moment index.

Reply:
Done.